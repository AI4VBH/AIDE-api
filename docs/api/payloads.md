<!--
  ~ Copyright 2022 Crown Copyright
  ~
  ~ Licensed under the Apache License, Version 2.0 (the "License");
  ~ you may not use this file except in compliance with the License.
  ~ You may obtain a copy of the License at
  ~
  ~ http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing, software
  ~ distributed under the License is distributed on an "AS IS" BASIS,
  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ~ See the License for the specific language governing permissions and
  ~ limitations under the License.
-->

# Payloads

The `payloads` endpoints provide the following APIs to retrieve data related to payloads.

## `GET /payloads?pageNumber=:pageNumber&pageSize=:pageSize`

Returns a PagedResponse list of payloads.

### Parameters

| Parameter Type | Label | Description | Format | Example |
|----------------|-------|-------------|--------|---------|
| Query | `:pageNumber` | `number` Page number for request. | 0-9 | 99
| Query | `:pageSize` | `number` Page size for request. | 0-9 | 99

### Responses

| Code | Description |
|------|-------------|
| `200` | PagedResponse of Payloads retrieved successfully. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request GET 'http://localhost:5000/payloads?pageNumber=1&pageSize=2'
```

### Example Response

```json
{
  "pageNumber": 1,
  "pageSize": 1,
  "firstPage": "/payload?pageNumber=1&pageSize=1",
  "lastPage": "/payload?pageNumber=1&pageSize=1",
  "totalPages": 100,
  "totalRecords": 100,
  "nextPage": 2,
  "previousPage": null,
  "data": [
            {
                "payload_id": "86c0f117-4021-412e-b163-0dc621df672a",
                "patient_id": "1d0253c4-8fab-41df-a414-55d52e4c6c3f",
                "patient_name": "Lillie Doe",
                "payload_received": "2022-08-17T12:21:10.203Z",
            }
          ],
}
```

---

## `GET /payloads/:payloadId/executions`

Returns a list of WorkflowInstances executed from payload.

### Parameters

| Parameter Type | Label | Description | Format | Example |
|----------------|-------|-------------|--------|---------|
| URL | `:payloadId` | `guid` of the payload. | GUID | 86c0f117-4021-412e-b163-0dc621df672a

### Responses

| Code | Description |
|------|-------------|
| `200` | Executions retrieved successfully. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request GET 'http://localhost:5000/payloads/86c0f117-4021-412e-b163-0dc621df672a/executions'
```

### Example Response

```json
[
    {
        "id": "c5e980ff-e278-441b-af64-8630bf57b7e3",
        "ae_title": "some-title",
        "payload_id": "041293d0-ab97-4ea1-b967-42ec62f26608",
        "workflow_id": "fca60a4d-2932-4196-9206-cf123c82332b",
        "start_time": "2022-09-22T12:45:13.102Z",
        "status": "Succeeded",
        "tasks": [
        {
            "task_id": "some-special-task",
            "workflow_instance_id": "e02e3bbc-a760-4d83-8cfe-ce3a77424171",
            "execution_id": "981eca65-294a-434a-930a-b1828b54253a",
            "previous_task_id": "",
            "status": "Suceeded",
            "task_start_time": "2022-09-23T12:50:00.305Z",
        },
        {
            "task_id": "export-task",
            "workflow_instance_id": "e02e3bbc-a760-4d83-8cfe-ce3a77424171",
            "execution_id": "981eca65-294a-434a-930a-b1828b54253a",
            "previous_task_id": "some-special-task",
            "status": "Suceeded",
            "task_start_time": "2022-09-23T12:50:00.305Z",
        },
        {
            "task_id": "email-task",
            "workflow_instance_id": "e02e3bbc-a760-4d83-8cfe-ce3a77424171",
            "execution_id": "981eca65-294a-434a-930a-b1828b54253a",
            "previous_task_id": "export-task",
            "status": "Suceeded",
            "task_start_time": "2022-09-23T12:50:00.305Z",
        },
        ],
    },
]
```