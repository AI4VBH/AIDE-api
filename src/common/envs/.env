PORT=5000

KEYCLOAK_URL='http://0.0.0.0:8443'
KEYCLOAK_MAIN_REALM='master'
KEYCLOAK_REALM='Demo-Realm'
KEYCLOAK_CLIENTID='nest-app'
KEYCLOAK_SECRET='83790b4f-48cd-4b6c-ac60-451a918be4b9'
KEYCLOAK_STATIC_ROLES='["admin", "clinician"]'

DATABASE_HOST='localhost'
DATABASE_NAME='nest_api'
DATABASE_USER='postgres'
DATABASE_PASSWORD='admin'
DATABASE_PORT=5432

ELASTIC_HOST=''
ELASTIC_PORT=''
ELASTIC_USE_SSL=''
ELASTIC_USERNAME=''
ELASTIC_PASSWORD=''
ELASTIC_INDEX=''

MONAI_API_HOST='https://localhost:8081/'
MIG_API_HOST='http://localhost:5010/'

MINIO_HOST='localhost'
MINIO_PORT=9001
MINIO_USE_SSL=false
MINIO_BUCKET='monaideploy'
MINIO_ACCESS_KEY='wOzmbchico52AMp3'
MINIO_SECRET_KEY='zyoT9SiMLiTCvNT0DS636FMiN4yi2258'
