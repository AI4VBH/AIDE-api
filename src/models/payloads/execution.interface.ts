export interface Execution {
  name: string;
}
