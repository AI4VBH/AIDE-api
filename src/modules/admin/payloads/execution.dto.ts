export class ExecutionDTO {
  execution_id: number;
  payload_id: number;
  model_name: string;
  execution_status: string;
  execution_started: string;
  execution_finished: string;
  executions: ExecutionDTO[];

  public static from(dto: Partial<ExecutionDTO>) {
    const it = new ExecutionDTO();
    it.execution_id = dto.execution_id;
    it.payload_id = dto.payload_id;
    it.model_name = dto.model_name;
    it.execution_status = dto.execution_status;
    it.execution_started = dto.execution_started;
    it.execution_finished = dto.execution_finished;
    it.executions = dto.executions;
    return it;
  }

  // TODO: Implement fromEntity method
  // public static fromEntity(entity: Execution) {
  //   return this.from({
  //     task_id: entity.id,
  //   });
  // }

  // TODO: Implement toEntity method
  // public toEntity(execution: Execution = null) {
  //   const it = new Execution();
  //   it.task_id = this.task_id;
  //   return it;
  // }
}
