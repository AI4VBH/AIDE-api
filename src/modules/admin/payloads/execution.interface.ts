export interface Execution {
  execution_id: number;
  payload_id: number;
  model_name: string;
  execution_status: string;
  execution_started: string;
  execution_finished: string;
  executions: Execution[];
}
