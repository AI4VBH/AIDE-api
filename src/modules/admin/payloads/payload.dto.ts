/*
 * Copyright 2022 Crown Copyright
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { IMonaiPayload } from './payload.interface';

export class PayloadDTO {
  payload_id: string;
  patient_name: string;
  patient_id: string;
  payload_received: string;

  public static from(dto: Partial<PayloadDTO>) {
    const it = new PayloadDTO();
    it.payload_id = dto.payload_id;
    it.patient_name = dto.patient_name;
    it.patient_id = dto.patient_id;
    it.payload_received = dto.payload_received;
    return it;
  }

  public static fromMonaiPayload(entity: IMonaiPayload) {
    return PayloadDTO.from({
      payload_id: entity.payload_id,
      patient_name: entity.patient_details.patient_name,
      patient_id: entity.patient_details.patient_id,
      payload_received: entity.timestamp,
    });
  }

  // TODO: Implement fromEntity method
  // public static fromEntity(entity: Payload) {
  //   return this.from({
  //     task_id: entity.id,
  //   });
  // }

  // TODO: Implement toEntity method
  // public toEntity(payload: Payload = null) {
  //   const it = new Payload();
  //   it.task_id = this.task_id;
  //   return it;
  // }
}
