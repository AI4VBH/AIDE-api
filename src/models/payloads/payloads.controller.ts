import { Controller, Get } from '@nestjs/common';
import { Execution } from './execution.interface';
import { Payload } from './payload.interface';
import { PayloadsService } from './payloads.service';

@Controller('payloads')
export class PayloadsController {
  constructor(private readonly appService: PayloadsService) {}

  @Get()
  getPayloads(): Payload {
    return this.appService.getPayloads();
  }

  @Get()
  getPayloadExecutions(): Execution {
    return this.appService.getPayloadExecutions();
  }
}
