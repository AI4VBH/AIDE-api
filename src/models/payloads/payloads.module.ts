import { Module } from '@nestjs/common';
import { PayloadsController } from './payloads.controller';
import { PayloadsService } from './payloads.service';

@Module({
  controllers: [PayloadsController],
  providers: [PayloadsService],
})
export class PayloadsModule {}
