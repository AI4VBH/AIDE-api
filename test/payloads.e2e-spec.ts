import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import { HttpService } from '@nestjs/axios';
import { ConfigService } from '@nestjs/config';
import * as request from 'supertest';
import { createMock, DeepMocked } from '@golevelup/ts-jest';
import { AxiosInstance } from 'axios';
import { PayloadsController } from 'modules/admin/payloads/payloads.controller';
import { PayloadsService } from 'modules/admin/payloads/payloads.service';
import { createObservableResponse } from './utilities/http-helpers';

describe('Payloads Controller', () => {
  let app: INestApplication;
  let axiosMock: DeepMocked<AxiosInstance>;
  let httpServiceMock: DeepMocked<HttpService>;
  let configServiceMock: DeepMocked<ConfigService>;

  beforeEach(async () => {
    axiosMock = createMock<AxiosInstance>();
    httpServiceMock = createMock<HttpService>({
      axiosRef: axiosMock,
    });
    configServiceMock = createMock<ConfigService>({
      get: (path) => path === 'MONAI_API_HOST' && 'http://localhost:7337',
    });

    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [],
      controllers: [PayloadsController],
      providers: [
        {
          provide: HttpService,
          useValue: httpServiceMock,
        },
        {
          provide: ConfigService,
          useValue: configServiceMock,
        },
        PayloadsService,
      ],
    })
      // .overrideProvider(HttpService)
      // .useValue(httpServiceMock)
      .compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });

  it('(GET) /payloads', () => {
    axiosMock.get.mockResolvedValue({ status: 404 });
    httpServiceMock.get.mockResolvedValue(
      createObservableResponse(axiosMock.get),
    );

    return request(app.getHttpServer())
      .get('/payloads?pageNumber=1&pageSize=10')
      .expect(200)
      .expect('test');
  });
});
