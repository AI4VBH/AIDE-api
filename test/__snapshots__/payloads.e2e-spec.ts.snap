// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payloads Controller (GET) /payloads with returned data 1`] = `
{
  "data": [
    {
      "patient_id": "1d0253c4-8fab-41df-a414-55d52e4c6c3f",
      "patient_name": "Jane Doe",
      "payload_id": "86c0f117-4021-412e-b163-0dc621df672a",
      "payload_received": "2022-08-17T12:21:10.203Z",
    },
    {
      "patient_id": "",
      "patient_name": "",
      "payload_id": "30a8e0c6-e6c4-458f-aa4d-b224b493d3c0",
      "payload_received": "2022-08-17T12:21:10.203Z",
    },
    {
      "patient_id": "fd1bebf4-d690-4fc5-a0d8-4fd4701ff4c9",
      "patient_name": "Steve Jobs",
      "payload_id": "c5c3636b-81dd-44a9-8c4b-71adec7d47b2",
      "payload_received": "2022-08-17T12:21:10.2Z",
    },
  ],
  "firstPage": "//payload?pageNumber=1&pageSize=10",
  "lastPage": "//payload?pageNumber=1&pageSize=10",
  "nextPage": null,
  "pageNumber": 1,
  "pageSize": 10,
  "previousPage": null,
  "totalPages": 1,
  "totalRecords": 3,
}
`;

exports[`Payloads Controller (GET) /payloads with returned data 2`] = `
{
  "message": "A decent error message",
  "statusCode": 400,
}
`;

exports[`Payloads Controller (GET) /payloads without returned data 1`] = `
{
  "data": [],
  "firstPage": "//payload?pageNumber=1&pageSize=10",
  "lastPage": "//payload?pageNumber=1&pageSize=10",
  "nextPage": null,
  "pageNumber": 1,
  "pageSize": 10,
  "previousPage": null,
  "totalPages": 1,
  "totalRecords": 3,
}
`;
